### ISSUE TITLE
Possibility to run without clickhouse/keycloak/prometheus?

### ISSUE THREAD

**Issue by jgautheron** [wundergraph/cosmo#90]
I've been playing with the local cosmo setup (which works great btw!), I got a few questions/concerns/suggestions - I don't know well the product so my assumptions could be wrong:
1. Possibility to run without clickhouse/prometheus (e.g. if the env vars are not found the metrics/tracing page is disabled in the control panel)
    - These two are quite expensive to run for production setups
    - We'd rather delegate metrics/tracing handling to Datadog
    - Additional risk/ services to monitor: if clickhouse/prometheus go down, then the control panel goes down as well?
2. Possibility to run the control panel without keycloak - similarly as above if the env vars are not found then the control panel works "anonymously", could save data in LocalStorage instead
    - Keycloak also quite expensive/complicated to run for production setups
    - Additional security concern since credentials are stored
3. My final concern is that the Go router depends on the control panel for fetching the subgraphs (maybe https://github.com/wundergraph/cosmo/pull/61 changes that?), the control panel depends on keycloak/postgres/clickhouse/prometheus, there's a lot of things that can go wrong scaling any of those - which could potentially take down the control panel, and therefore the router?
    - For production setups, having the possibility to deploy only the router without the control panel, would be great, only thing to scale, no database dependency

Keep up the good work!

**Comment by jgautheron** [comment]
> Just to point this out. For on-premise production, we recommend managed services as mentioned in https://github.com/wundergraph/cosmo/blob/main/helm/cosmo/README.md#production-use. Alternatively, you can join our Cloud and don't need to worry about anything 😄

Yes and that's good advice, but it does inflate the bill quite a bit, plus a few more third parties to deal with.

> What are your requirements here? A single admin user?

Could be either anonymous access or single admin user, a feature flag that would not require an idp.

> In all cases, the router will work without any of those components. Today, you can run a router with a static router configuration with [wgc router fetch](https://cosmo-docs.wundergraph.com/cli/router/fetch) without ever needing a connection to the controlplane. You can also point to your datadog agent for OLTP data. Metrics are exposed through prometheus endpoint and can be scraped from your infrastructure too.

That's great, I'll try that!

> Also, the controlplane was built in a way that it only requires postgres to function as the schema registry. An interruption of any other component should not result in downtime of the registry. Outlook: As illustrated here https://cosmo-docs.wundergraph.com/architecture the router will not fetch the config from the controlplane but from CDN. This will be behind a feature flag.

I see an arrow towards the control plane for "management". I suppose not a mandatory connection?

> Thank you! Are you speaking for a company? If yes, we can set up a call and talk about all your concerns in detail.

Yes we're currently looking to get into GraphQL federation; features we're looking for:
1. Preferably support for federation v2 (a bit confusing, sometimes Federation v2 is mentioned but here it says that v2 is not yet supported https://cosmo-docs.wundergraph.com/router/intro)
2. Subscriptions via the router (I didn't see any mention of that in the docs, don't know if it's supported)
3. JWT auth handled in the router (I don't think you support this out of the box but can be implemented with a custom go module)
4. Distributed caching (I don't think you support this out of the box but can be implemented with a custom go module)
5. Metrics/tracing to Datadog (YES with OLTP/ pulling metrics from the prometheus endpoint)
6. Static schema (not sure if you support it - https://cosmo-docs.wundergraph.com/cli/router/compose looks dynamic, I don't see a local supergraph.gql file mentioned)

**Comment by Aenimus** [comment]
@jgautheron 
Regarding Federation V2, it is indeed supported. There are some V2 directives that are currently in development, but the chances are you're not/wouldn't be using those.

Regarding this question and your other questions, it might be worth a more in-depth discussion/explanation. Is there an email by which I can reach you? You could also send the email to me privately on Twitter: https://twitter.com/aenimuskol

**Comment by Aenimus** [comment]
@jgautheron I saw you have followed me on Twitter. I sent you a message, which you can access here: https://twitter.com/messages

Speak soon 🚀

**Comment by StarpTech** [comment]
> See an arrow towards the control plane for "management". I suppose not a mandatory connection?

Correct, not mandatory.

> Subscriptions via the router (I didn't see any mention of that in the docs, don't know if it's supported)

Will be released very soon.

> Distributed caching (I don't think you support this out of the box but can be implemented with a custom go module)

That's correct.

> Preferably static schema (not sure if you support it - [cosmo-docs.wundergraph.com/cli/router/compose](https://cosmo-docs.wundergraph.com/cli/router/compose) looks dynamic, I don't see a local supergraph.gql file mentioned)

The compose command exists to allow local composition without the control-plane. It produces a static router config which is the equivalent of the supergraph.gql.I think you're looking for is https://cosmo-docs.wundergraph.com/cli/router/fetch You can fetch the router config once and use it to start the router.

> Multiple routers (e.g one public, one private) - probably no problem if config embedded, that looks also possible/doable if connected to the control plane since the router has the config graph/name.

That would be the right and common approach.
